### PR TITLE
Create a code analyzer to avoid using `async void` #2766

### LINKED ISSUE #2766: Code Analyzer to avoid using `async void`
Based on #2765, it is very important to have a code analyzer to warn the developers about `async void` usages.

### PR DESCRIPTION
**Pull request by milad-karimifard**
this closes #2766

### CODE REVIEW

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        if (returnType is null)
+            return;
+
+        if (returnType.Keyword.IsKind(SyntaxKind.VoidKeyword) is false)
+            return;
+
+        if (methodDec.Body is null)
+            return;
+
+        if (methodDec.Body.Statements.Any(SyntaxKind.TryStatement))
```
Comment (ysmoradi) [inline src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs:58]: if (methodDec.Body?.Statements?.Any(SyntaxKind.TryStatement) is true)
            return;

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        var returnType = methodDec.ReturnType as PredefinedTypeSyntax;
+
+        if (returnType is null)
+            return;
+
+        if (returnType.Keyword.IsKind(SyntaxKind.VoidKeyword) is false)
+            return;
+
+        if (methodDec.Body is null)
```
Comment (ysmoradi) [inline src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs:55]: remove this

**Review by ysmoradi — CHANGES_REQUESTED** [review]

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace Bit.CodeAnalyzers.SystemAnalyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class MethodAsyncVoidAnalyzer : DiagnosticAnalyzer
+{
+    public const string DiagnosticId = nameof(MethodAsyncVoidAnalyzer);
+
+    public const string Title = "Change return type of async method to Task";
+    public const string Message = "Return type of async method must be Task or contain try/catch";
```
Comment (ysmoradi) [inline src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs:16]: Could you please suggest a better message? @msynk

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs (hunk 1 above)
Comment (msynk) [reply src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs:58]: `is true` ???

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs (hunk 3 above)
Comment (msynk) [inline src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/MethodAsyncVoidAnalyzer.cs:16]: please change this value to this:
`Either change the return type to 'Task' or use try/catch to handle the exceptions.`

**Review by msynk — CHANGES_REQUESTED** [review]

**Review by msynk — DISMISSED** [review]

**Review by ysmoradi — DISMISSED** [review]

File: src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/AsyncVoidMethodAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if (returnType.Keyword.IsKind(SyntaxKind.VoidKeyword) is false)
+            return;
+
+        if (methodDec.Body?.Statements.FirstOrDefault()?.IsKind(SyntaxKind.TryStatement) is false)
+        {
+            RepoertError(context, root, DefaultRule);
+            return;
+        }
+
+        if (IsTryStatementUsedCorrectly(methodDec) is true)
```
Comment (Cyrus-Sushiant) [inline src/CodeAnalyzers/Bit.CodeAnalyzers/SystemAnalyzers/AsyncVoidMethodAnalyzer.cs:66]: plz, remove `is true`

**Review by Cyrus-Sushiant — CHANGES_REQUESTED** [review]

**Review by Cyrus-Sushiant — APPROVED** [review]

**Review by msynk — APPROVED** [review]

**Pull request merged**